File: Python/398. Random Pick Index.py
class Solution:

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.nums = nums
        

    def pick(self, target):
        """
        :type target: int
        :rtype: int
        """
        cnt, ans = 0, 0
        for i in range(len(self.nums)):
            if target != self.nums[i]:
                continue
            cnt += 1
            if cnt == 1:
                ans = i
            elif random.randint(1, cnt) == cnt:
                    ans = i
        return ans


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.pick(target)

File: C++/398. Random Pick Index.cpp
class Solution {
    vector<int> _nums;
public:
    Solution(vector<int> nums) {
        _nums = nums;
    }
    
    int pick(int target) {
        int cnt = 0, ans = 0;
        for(int i=0;i<_nums.size();i++)
        {
            if(target != _nums[i])
                continue;
            cnt += 1;
            if(cnt == 1)
                ans = i;
            else
            {
                if(!(rand()%cnt))
                    ans = i;
            }
        }
        return ans;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int param_1 = obj.pick(target);
 * 水塘抽样
 */